agent/php_user_instrument.c

```diff
@@ -395,7 +395,8 @@ nruserfn_t* nr_php_add_custom_tracer_callable(zend_function* func TSRMLS_DC) {
 }
 
 nruserfn_t* nr_php_add_custom_tracer_named(const char* namestr,
-                                           size_t namestrlen TSRMLS_DC) {
+                                           size_t namestrlen,
+                                           bool is_transient TSRMLS_DC) {
```

Comment:
To minimize code changes and keep backwards compatibility this code should be:
```
nruserfn_t* nr_php_add_custom_tracer_named(const char* namestr,
                                           size_t namestrlen TSRMLS_DC) {

  nr_php_add_custom_tracer_named_with_transience(namestr, namestrlen, false, TSRMLS_CC);
}

nruserfn_t* nr_php_add_custom_tracer_named_with_transience(const char* namestr,
                                           size_t namestrlen,
                                           bool is_transient TSRMLS_DC) {
...
```
As in:
1. Leave behavior of `nr_php_add_custom_tracer_named` unchanged, which is never create transient wraprec
2. Add a new function `nr_php_add_custom_tracer_named_with_transience` that will add new behavior, i.e. it can create a wraprec that can be transient or not depending on the input (the value of new argument).
This way no place in the code will require change except for the place that requires transient wraprec for `custom_tracer_named` to be added.

Reply:
I had considered this approach but ultimately decided against it. This complicates the API by adding yet another option to call. Because the `add_custom_tracer` type functions are quite limited in where they used, I do not think it is justified to add confusion with this interface (unlike with the `wrap_user_function` family which is invoked in MANY locations, and 99% of these calls are with is_transient=false). Ultimately when reasonable, I would always rather have a function with more parameters than multiple functions used to emulate default parameters (thanks C for not having those).

The side-effects here is that in the future if we ever add a place calling nr_php_add_custom_tracer_named, we will be forced to consider whether each call should be transient- making us less likely to make the same mistake that caused this PR.

Reply:
[This approach](https://github.com/newrelic/newrelic-php-agent/pull/679#discussion_r1228571233) does not complicate things - it simplifies things and removes the `/* is_transient */` comment from the code which eventually will become confusing and dead.

Reply:
Something like this so when looking at code it is obvious what type is being created would be nice.  Does this get rid of may of the /* is_transient */ in the code?

Reply:
But what if in the future we want to add another option (similar to this one) to this interface? Even just 1 more option here would immediately balloon into 4 more function headers for the same underlying call, quickly becoming unwieldy and unreadable.

Reply:
> Something like this so when looking at code it is obvious what type is being created would be nice. Does this get rid of may of the /* is_transient */ in the code?

Yes. Both suggestions eliminate `/* is_transient */` comments in the code.

Reply:
> Having two APIs with `_transient` and `_persistent` suffixes makes it clear what the API will do and what caller intention was which makes it precisely the case in point - the author calling this API knows what they're doing.

If you are reading the API/header, yes. If one is just reading the code and see the call somewhere, then to get this info one would need to code dive.

Reply:
refactored in [ea328b5](https://github.com/newrelic/newrelic-php-agent/pull/679/commits/ea328b5b408d4efa889f4e6bc9d2733cca2301e2)

Reply:
Thank you for compromising.